File: Project/Discipline.cpp
#include "Discipline.hpp"

#include "StringUtils.hpp"

#include <stdexcept>


Discipline::Discipline(std::string name, AttestationType attestationType) {
	this->SetName(name);
	m_attestationType = attestationType;
}

const std::string& Discipline::GetName() const {
	return m_name;
}

void Discipline::SetName(const std::string& value) {
	std::string newValue = StringUtils::Trim(value);
	if (newValue.size() == 0) {
		throw std::invalid_argument("Discipline name cannot be empty");
	}
	m_name = std::move(newValue);
}

Discipline::AttestationType Discipline::GetAttestationType() const {
	return m_attestationType;
}

void Discipline::SetAttestationType(Discipline::AttestationType value) {
	m_attestationType = value;
}

bool Discipline::Equals(const Discipline& other) const {
	return m_name == other.GetName() && m_attestationType == other.GetAttestationType();
}

File: ProjectCS/Data/Discipline.cs
using System;

namespace SessionResultsCS
{
    // Перечисление типов аттестации
    public enum AttestationType
    {
        Exam,           // Экзамен
        PassFailExam    // Зачёт
    }

    public class Discipline
    {
        private string _name;


        // Название дисциплины
        public string Name {
            get { return _name; }
            set
            {
                string newName = value.Trim();
                if (newName.Length == 0)
                {
                    throw new ArgumentException("Discipline name cannot be empty");
                }
                _name = newName;
            }
        }

        // Тип аттестации
        public AttestationType AttestationType { get; set; }


        public Discipline(string name, AttestationType attestationType)
        {
            Name = name;
            AttestationType = attestationType;
        }


        public Discipline Clone()
        {
            return new Discipline(Name, AttestationType);
        }


        // Проверяет равенство с другой дисциплин
        public override bool Equals(object? obj)
        {
            if (obj is Discipline other)
            {
                return this == other;
            }
            return false;
        }


        public static bool operator ==(Discipline left, Discipline right)
        {
            return left.Name == right.Name && left.AttestationType == right.AttestationType;
        }

        public static bool operator !=(Discipline left, Discipline right)
        {
            return !(left == right);
        }
    }
}
